core/arch/arm/include/mm/core_mmu.h

```diff
@@ -165,22 +165,56 @@ static inline const char *teecore_memtype_name(enum teecore_memtypes type)
 struct core_mmu_phys_mem {
 	const char *name;
 	enum teecore_memtypes type;
-	paddr_t addr;
-	size_t size;
+	__extension__ union {
+#if __SIZEOF_LONG__ != __SIZEOF_PADDR__
+		struct {
+			uint32_t lo_addr;
+			uint32_t hi_addr;
+		};
+#endif
+		paddr_t addr;
+	};
+	__extension__ union {
+#if __SIZEOF_LONG__ != __SIZEOF_PADDR__
+		struct {
+			uint32_t lo_size;
+			uint32_t hi_size;
+		};
+#endif
+		paddr_size_t size;
+	};
 };
 
 #define __register_memory2(_name, _type, _addr, _size, _section, _id) \
 	static const struct core_mmu_phys_mem __phys_mem_ ## _id \
 		__used __section(_section) = \
 		{ .name = _name, .type = _type, .addr = _addr, .size = _size }
 
+#if __SIZEOF_LONG__ != __SIZEOF_PADDR__
```

Comment:
looks strange to have 2 macros `register_phys_mem()` and `register_phys_mem_ul()`.
Can't use something like:
```
#if __SIZEOF_LONG__ == __SIZEOF_PADDR__
#define __PHYSMEM_LOC(_addr, _size) \
			.addr = (_addr), .size = (_size),
#else
#define __PHYSMEM_LOC(_addr, _size) \
			.lo_addr = (_addr), .lo_size = (_size),
#endif

#define __register_memory2(_name, _type, _addr, _size, _section, _id) \
	static const struct core_mmu_phys_mem __phys_mem_ ## _id \
		__used __section(_section) = \
 		{ .name = _name, .type = _type, __PHYSMEM_LOC(_addr, _size) },
```

Reply:
We need both the normal and the _ul version. The normal is needed when supplying 64-bit addresses from a proper constant, while the _ul version must be used when initializing from variables provided by the link script.

It's not beautiful and have some sharp corners, but I can't find a way around it.